File: InterogateNGRAM.py
#! /usr/bin/python

import MicrosoftNgram
import sys, io, os, time

E_VALUE = 2.71828
MAX_LENGTH = 55000

lookup = MicrosoftNgram.LookupService()

fdreq = int(sys.argv[1])
fdres = int(sys.argv[2])
#print 'py am primit fd-urile pipurilor anonime: ' + str(fdreq) + ' ' + str(fdres)

requestInput = os.fdopen(fdreq, "r")
resultOutput = os.fdopen(fdres, "w")
#print 'py am deschis fd-urile'

out_deb = open("InterogateNGRAM.out", "w", 0)
out_deb.write('am deschis iesirea\n')
#print 'py am deschis iesirea\n'

buffer = []
buffering = False
buffer_len = 0
result = ""

while (1):
    #print 'py astept sa primesc ceva de la requestInput'
    line = requestInput.readline()
    out_deb.write('am citit ceva de la requestInput\n')
    #print 'py am citit ceva de la requestInput\n'
    if line:
        out_deb.write('am primit ' + line)
        #print 'py am primit ' + line
        if (line == 'Interogate please exit\n'):
            requestInput.close()
            resultOutput.close()
            exit(0)
        elif (line == 'Interogate please start buffering\n'):
            out_deb.write('Incep sa fac buffer\n')
            buffer = []
            buffering = True
            buffer_len = 0
            result = ""
            continue
        elif (line == 'Interogate please send queries\n'):
            #out_deb.write('Am terminat de facut buffer\n')
            """
            Check if we have previous results
            """
            if result != "":
                result += '\n'
            if buffer != []:
                result += lookup.GetJointProbabilities(buffer)
            result_final = ""
            for nr in result.split('\n'):
                result_final = result_final + str(E_VALUE**(float(nr))) + "\n"
            #out_deb.write("result " + result_final)
            resultOutput.write(result_final)
            resultOutput.flush()
            #out_deb.write('am trimis rezultat final\n')
            buffer = []
            buffering = False
            buffer_len = 0
            result = ""
            continue

        if buffering:
            '''remove the '\n' from the end of the line'''
            buffer.append(line[:-1])
            buffer_len += len(line[:-1])
            #out_deb.write("dimensiunea bufferului: " + str(buffer_len) + "\n")
            if buffer_len > MAX_LENGTH:
                out_deb.write("am atins dimensiunea maxima, fac req\n")
                """
                Reached maximum size of html requests
                """
                if result != "":
                    result += '\n'
                result += lookup.GetJointProbabilities(buffer)
                buffer = []
                buffer_len = 0
            continue
        result = str(E_VALUE**lookup.GetJointProbability(line))
        #out_deb.write('am calculat rezultatul ' + result + ' pentru ' + line + '\n')
        resultOutput.write(result + '\n')
        resultOutput.flush()
    else:
        time.sleep(1)


requestInput.close()
resultOutput.close()

File: InterogateNGRAM.cpp
#include "InterogateNGRAM.h"

#include <vector>
#include <string>
#include <iostream>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

//mkfifo
#include <sys/types.h>
#include <sys/stat.h>
using namespace std;

//if you run this program on a very busy machine
//consider incresing this value
#define FD_SIZE 5

FILE* InterogateNGRAM::resultPipe;
FILE* InterogateNGRAM::requestPipe;

float InterogateNGRAM::getJointProbability(std::vector<std::string> &phrase)
{
    std::string strPhrase = "";
    for(unsigned int i = 0; i < phrase.size() - 1; i++)
    {
        strPhrase += phrase[i] + " ";
    }
    strPhrase += phrase[phrase.size() - 1];

    return InterogateNGRAM::getJointProbability(strPhrase);
}

float InterogateNGRAM::getJointProbability(std::string phrase)
{
    //trimit request
    phrase += "\n";
    fputs(phrase.c_str(), requestPipe);
    fflush(requestPipe);

    //primesc rezultatul
    char buffer[20];
    fgets(buffer, 20, resultPipe);
    //puts(buffer);

    float result = atof(buffer);
    return result;
}

std::vector<float> InterogateNGRAM::getJointProbabilities(std::vector<std::string> &phrases)
{
    std::vector<float> result;
    char buffer[100];

    /* Protocol:
     * send a unique string: "Interogate please start buffering\n"
     * send all ngrams separated by "\n"
     * send a unique string: "Interogate please send queries"
     */
    fputs("Interogate please start buffering\n", requestPipe);
    //printf("am trimis interogate pls start\n");
    for(unsigned int i = 0; i < phrases.size(); i++)
    {
        strcpy(buffer, (phrases[i] + "\n").c_str());
        //printf("sending %s\n", buffer);
        fputs(buffer, requestPipe);
        //
        //  ATTENTION
        //
        //if InterogateNGRAM.py locks on readline
        //you should decomment this line
        //fflush(requestPipe);
    }
    fputs("Interogate please send queries\n", requestPipe);
    fflush(requestPipe);
    //printf("am trimis interogate pls send\n");

    result.clear();
    for(unsigned int i = 0; i < phrases.size(); i++)
    {
        //printf("incep fgets\n");
        fgets(buffer, 100, resultPipe);
        //printf("termin fgets\n");
        //printf("deb %s\n", buffer);
        result.push_back(atof(buffer));
    }

    return result;
}

void InterogateNGRAM::Init()
{
    //printf("init 1\n");
    //mkfifo("/tmp/ngramfifo", 0666);
    //mkfifo("/tmp/ngramfiforeq", 0666);

    //unnamed pipes
    //each pipe is: input | output
    int requestPipeFd[2], resultPipeFd[2];
    if(pipe(requestPipeFd) == -1)
    {
        printf("ERROR creating the annonymus request pipe\n");
        exit(0);
    }
    if(pipe(resultPipeFd) == -1)
    {
        printf("ERROR creating the annonymus result pipe\n");
        exit(0);
    }
    //printf("pipe: am creat pipe-urile: %d %d %d %d\n", requestPipeFd[0], requestPipeFd[1], resultPipeFd[0], resultPipeFd[1]);

    pid_t pid = fork();
    if(pid == -1)
    {
        printf("ERROR starting the InterogateNGRAM.py process\n");
        exit(0);
    }
    if(pid == 0)
    {
        //child process
        close(requestPipeFd[1]);
        close(resultPipeFd[0]);
        char requestPipeFdStr[FD_SIZE], resultPipeFdStr[FD_SIZE];
        sprintf(requestPipeFdStr, "%d", requestPipeFd[0]);
        sprintf(resultPipeFdStr, "%d", resultPipeFd[1]);
        int ret = execlp("python", "python", "InterogateNGRAM.py", requestPipeFdStr, resultPipeFdStr, NULL);
        if(ret == -1)
        {
            printf("ERROR at execlp\n");
        }
    }
    //parent process
    //printf("init 2\n");
    close(requestPipeFd[0]);
    close(resultPipeFd[1]);
    //requestPipe = fopen("/tmp/ngramfiforeq", "w");
    requestPipe = fdopen(requestPipeFd[1], "w");
    if(requestPipe == NULL)
    {
        printf("ERROR opening request pipe\n");
    }
    //printf("init 3\n");
    //resultPipe = fopen("/tmp/ngramfifo", "r");
    resultPipe = fdopen(resultPipeFd[0], "r");
    if(resultPipe == NULL)
    {
        printf("ERROR opening result pipe\n");
    }
    //printf("init 4\n");
}

void InterogateNGRAM::Finalize()
{
    fputs("Interogate please exit\n", requestPipe);
    fclose(resultPipe);
    fclose(requestPipe);
}
